File: server/proto.go
// SPDX-License-Identifier: GPL-3.0-or-later
//
// Server side of the Ur wire protocol. Must stay byte-for-byte identical to
// src/common/proto.{h,c} and docs/protocol.md.
package main

const (
	MsgJoin  = 0x01 // client->server: join (byte1 = version)
	MsgRoll  = 0x02 // client->server: request a roll
	MsgMove  = 0x03 // client->server: move (byte1 = piece index)
	MsgState = 0x81 // server->client: 21-byte snapshot
)

const (
	PhaseRoll = 0
	PhaseMove = 1
	PhaseOver = 2
)

const (
	FlagCaptured = 0x01
	FlagRosette  = 0x02
	FlagScored   = 0x04
)

const (
	ProtoVersion = 1
	StateMsgLen  = 21
)

// encodeState builds the 21-byte STATE snapshot for a given seat.
func encodeState(seat, phase, roll uint8, winner int, flags uint8, st *State) []byte {
	b := make([]byte, StateMsgLen)
	b[0] = MsgState
	b[1] = seat
	b[2] = st.Turn
	b[3] = phase
	b[4] = roll
	if winner < 0 {
		b[5] = 0xFF
	} else {
		b[5] = uint8(winner)
	}
	b[6] = flags
	for i := 0; i < Pieces; i++ {
		b[7+i] = st.Piece[0][i]
		b[14+i] = st.Piece[1][i]
	}
	return b
}

func flagsFrom(r MoveResult) uint8 {
	var f uint8
	if r.Captured {
		f |= FlagCaptured
	}
	if r.Rosette {
		f |= FlagRosette
	}
	if r.Scored {
		f |= FlagScored
	}
	return f
}

File: src/common/proto.c
/* SPDX-License-Identifier: GPL-3.0-or-later */
/* Ur multiplayer wire-protocol codec. See proto.h and docs/protocol.md. */

#include "proto.h"

uint8_t ur_proto_join(uint8_t *buf)
{
    buf[0] = UR_MSG_JOIN;
    buf[1] = UR_PROTO_VERSION;
    return 2;
}

uint8_t ur_proto_roll(uint8_t *buf)
{
    buf[0] = UR_MSG_ROLL;
    return 1;
}

uint8_t ur_proto_move(uint8_t *buf, uint8_t piece)
{
    buf[0] = UR_MSG_MOVE;
    buf[1] = piece;
    return 2;
}

uint8_t ur_proto_encode_state(uint8_t *buf, const ur_snapshot *snap)
{
    uint8_t i, n = 0;

    buf[n++] = UR_MSG_STATE;
    buf[n++] = snap->seat;
    buf[n++] = snap->state.turn;
    buf[n++] = snap->phase;
    buf[n++] = snap->roll;
    buf[n++] = (uint8_t)(snap->winner < 0 ? 0xFF : (uint8_t)snap->winner);
    buf[n++] = snap->flags;
    for (i = 0; i < UR_PIECES; i++)
        buf[n++] = snap->state.piece[0][i];
    for (i = 0; i < UR_PIECES; i++)
        buf[n++] = snap->state.piece[1][i];
    return n;                       /* == UR_STATE_MSG_LEN */
}

bool ur_proto_decode_state(const uint8_t *buf, uint8_t len, ur_snapshot *snap)
{
    uint8_t i, n;
    uint8_t w;

    if (len < UR_STATE_MSG_LEN || buf[0] != UR_MSG_STATE)
        return false;

    n = 1;
    snap->seat       = buf[n++];
    snap->state.turn = buf[n++];
    snap->phase      = buf[n++];
    snap->roll       = buf[n++];
    w                = buf[n++];
    snap->winner     = (w == 0xFF) ? (int8_t)-1 : (int8_t)w;
    snap->flags      = buf[n++];
    for (i = 0; i < UR_PIECES; i++)
        snap->state.piece[0][i] = buf[n++];
    for (i = 0; i < UR_PIECES; i++)
        snap->state.piece[1][i] = buf[n++];
    return true;
}
